### ISSUE TITLE
Windows 10 exe file detail info has messy code text

### ISSUE THREAD

**Issue by ClayGminx** [fvarrui/JavaPackager#325]
I'm submitting a…

- [x] bug report
- [ ] feature request
- [ ] other

As the following picture display, the detail info in exe file has messy code text.

Note that my true description is "一个用于自动制作敬拜PPT的小工具".

Additional issue, my expect language is "简体中文", but it display "英语(美国)".

Following is my pom.xml

```
<name>Worship PPT</name>
<description>一个用于自动制作敬拜PPT的小工具</description>
......
<plugin>
    <groupId>io.github.fvarrui</groupId>
    <artifactId>javapackager</artifactId>
    <version>1.7.2</version>
    <executions>
        <execution>
            <phase>package</phase>
            <goals>
                <goal>package</goal>
            </goals>
            <configuration>
                <mainClass>claygminx.worshipppt.Main</mainClass>
                <bundleJre>true</bundleJre>
                <customizedJre>true</customizedJre>
                <jdkPath>C:\Program Files\Java\jdk1.8.0_351</jdkPath>
                <jrePath>jre</jrePath>
                <generateInstaller>false</generateInstaller>
                <administratorRequired>false</administratorRequired>
                <platform>auto</platform>
                <createZipball>true</createZipball>
                <createTarball>true</createTarball>
                <url>https://github.com/ClayGminx/worship-ppt</url>
                <additionalResources>
                    <additionalResource>config</additionalResource>
                    <additionalResource>PPT模板</additionalResource>
                    <additionalResource>数据库</additionalResource>
                    <additionalResource>README.md</additionalResource>
                </additionalResources>

                <winConfig>
                    <copyright>Joseph Mark</copyright>
                    <language>zh_CN</language>
                </winConfig>
            </configuration>
        </execution>
    </executions>
</plugin>
```

And following is the exe file detail info.

![Worship PPT exe issue](https://github.com/fvarrui/JavaPackager/assets/62022018/061e4d7c-4622-4238-9a2f-93f7a73cc50b)

**Comment by fvarrui** [comment]
Hi @ClayGminx!
Please, take a look to issue #21. We haven't found an automated solution to this problem 😞

**Comment by fvarrui** [comment]
Hi @ClayGminx!
I've been researching a bit about this issue.

Please, could you try setting `winConfig.language=SIMPLIFIED_CHINESE` instead of `zh_CN`?

```xml
<plugin>
    <groupId>io.github.fvarrui</groupId>
    <artifactId>javapackager</artifactId>
    <version>1.7.2</version>
    <executions>
        <execution>
            <phase>package</phase>
            <goals>
                <goal>package</goal>
            </goals>
            <configuration>
                <mainClass>claygminx.worshipppt.Main</mainClass>
                <bundleJre>true</bundleJre>
                <customizedJre>true</customizedJre>
                <jdkPath>C:\Program Files\Java\jdk1.8.0_351</jdkPath>
                <jrePath>jre</jrePath>
                <generateInstaller>false</generateInstaller>
                <administratorRequired>false</administratorRequired>
                <platform>auto</platform>
                <createZipball>true</createZipball>
                <createTarball>true</createTarball>
                <url>https://github.com/ClayGminx/worship-ppt</url>
                <additionalResources>
                    <additionalResource>config</additionalResource>
                    <additionalResource>PPT模板</additionalResource>
                    <additionalResource>数据库</additionalResource>
                    <additionalResource>README.md</additionalResource>
                </additionalResources>

                <winConfig>
                    <copyright>Joseph Mark</copyright>
                    <language>SIMPLIFIED_CHINESE</language>
                </winConfig>
            </configuration>
        </execution>
    </executions>
</plugin>
```

**Comment by lhDream** [comment]
@fvarrui 
See #202 ResourceHacker.exe 
How about a built-in ResourceHacker.exe ?

**Comment by lhDream** [comment]
![image](https://github.com/fvarrui/JavaPackager/assets/41568266/044f8949-476d-4a98-b1bc-f2e48bf5fd8d)
configuration used :
```xml
<plugin>
    <groupId>io.github.fvarrui</groupId>
    <artifactId>javapackager</artifactId>
    <version>1.7.2</version>
    <executions>
        <execution>
            <phase>package</phase>
            <goals>
                <goal>package</goal>
            </goals>
            <configuration>
                <displayName>中文测试</displayName>
                <mainClass>org.example.SignTest</mainClass>
                <bundleJre>true</bundleJre>
                <customizedJre>true</customizedJre>
                <jdkPath>C:\Program Files\Java\jdk1.8.0_171</jdkPath>
                <jrePath>${java.home}</jrePath>
                <generateInstaller>false</generateInstaller>
                <administratorRequired>false</administratorRequired>
                <platform>auto</platform>
                <createZipball>true</createZipball>
                <createTarball>true</createTarball>
                <url>https://github.com/ClayGminx/worship-ppt</url>
                <additionalResources>
                    <additionalResource>config</additionalResource>
                    <additionalResource>PPT模板</additionalResource>
                    <additionalResource>数据库</additionalResource>
                    <additionalResource>README.md</additionalResource>
                </additionalResources>

                <winConfig>
                    <copyright>Joseph Mark</copyright>
                    <language>SIMPLIFIED_CHINESE</language>
                </winConfig>
            </configuration>
        </execution>
    </executions>
</plugin>
```

**Comment by fvarrui** [comment]
hi @lhDream!
Thanks for testing it

**Comment by ClayGminx** [comment]
Maybe due to the encoding of my pom.xml is UTF-8, but my win10 encoding is GBK. I test it usding SIMPLIFIED_CHINESE, but still messy code.

**Comment by ClayGminx** [comment]
If due to different encoding, can you transform it?

**Comment by fvarrui** [comment]
> @fvarrui See #202 ResourceHacker.exe How about a built-in ResourceHacker.exe ?

![image](https://github.com/fvarrui/JavaPackager/assets/4354486/0308e981-22c6-46b3-93da-c4684e163850)

We can't distribute RH with JavaPackager, but maybe JP can use it if the user pre-installs it, like it's done with Inno Setup, for example.

Right now JP is using [`rcedit`](https://github.com/electron/rcedit) which is under MIT license.

According to RH docs, it can be used from command line with its own scripting language:

```bash
ResourceHacker.exe -script ScriptFile
```

**Comment by fvarrui** [comment]
After some testing I've realized that you can use `winconfig.exeCreationTool=why|winrun4j` ... these options set EXE file description fine.

**Comment by ClayGminx** [comment]
Hello! When I use
```xml
<exeCreationTool>why</exeCreationTool>
```
The package result is the following
![2](https://github.com/fvarrui/JavaPackager/assets/62022018/9b15ec76-5958-4e78-80ee-443eb77afec6)
I can't use my app normally.

And about Launch4j GUI, I have tried it, but I have not focus the detail info.

**Comment by fvarrui** [comment]
Hi @ClayGminx!

I've been doing some tests with your project and I think I finally found a solution. I don't know why `Why` doesn't work for your application, but I got it to work using `WinRun4J`.

I also had to improve several things in JP, because some metadata was not being set in the EXE file when using `why` and `winrun4j`, and I published a SNAPSHOT version with the changes.

Set `winConfig.exeCreationTool=winrun4j` and use this JP SNAPSHOT version: `1.7.3-20230613.105302-8`. You should also add the following snippet to your POM, so that Maven can find snapshots:

```xml
	<pluginRepositories>
		<pluginRepository>
			<id>nexus</id>
			<name>nexus-snapshot-repository</name>
			<url>https://oss.sonatype.org/content/repositories/snapshots</url>
			<snapshots>
				<enabled>true</enabled>
				<updatePolicy>always</updatePolicy>
			</snapshots>
			<releases>
				<enabled>false</enabled>
			</releases>
		</pluginRepository>
	</pluginRepositories>
```

Please, try it and give me some feedback. Thanks!

**Comment by fvarrui** [comment]
I forgot it:

```xml
                    <execution>
                        <id>bundling-for-windows</id>
                        <phase>package</phase>
                        <goals>
                            <goal>package</goal>
                        </goals>
                        <configuration>
                            <platform>windows</platform>
                            <createZipball>false</createZipball>
                            <createTarball>false</createTarball>
                            <jrePath>C:\Program Files\Java\jdk1.8.0_211\jre</jrePath>
                            <winConfig>
                                <exeCreationTool>winrun4j</exeCreationTool>
                                <!--
                                <fileDescription>A worship PPT Tool</fileDescription>
                                -->
                                <copyright>Joseph Mark</copyright>
                            </winConfig>
                        </configuration>
                    </execution>
```

If you don't specify `winConfig.fileDescription`, JP uses the POM's `description` by default, and if you are bundling a JRE 8, you don't have to use `jdkPath` property, as this is only when you want to generate a customized JRE (only JDK 9+). You only have to specify `jrePath` property, pointing to the JRE you want to bundle with your app.

You can also download and unzip a JRE for MacOS on Windows and set `platform=mac` and `jrePath=path/to/macos/jre`... this will allow you to build the Mac OS application from Windows.

**Comment by ClayGminx** [comment]
@fvarrui thank you. I will try it soonly.

**Comment by ClayGminx** [comment]
@fvarrui Hello! I tried it. I used the following pom.xml

```xml
<build>
    <finalName>${project.name}-${project.version}</finalName>
    <resources>
        <resource>
            <directory>src/main/resources</directory>
            <filtering>true</filtering>
            <includes>
                <include>*.*</include>
                <include>templates/*.*</include>
            </includes>
        </resource>
    </resources>
    <plugins>
        <plugin>
            <groupId>org.apache.maven.plugins</groupId>
            <artifactId>maven-resources-plugin</artifactId>
            <version>3.1.0</version>
            <configuration>
                <encoding>UTF-8</encoding>
                <nonFilteredFileExtensions>
                    <nonFilteredFileExtension>pptx</nonFilteredFileExtension>
                    <nonFilteredFileExtension>db</nonFilteredFileExtension>
                </nonFilteredFileExtensions>
            </configuration>
        </plugin>
        <plugin>
            <groupId>org.apache.maven.plugins</groupId>
            <artifactId>maven-compiler-plugin</artifactId>
            <version>3.5.1</version>
            <configuration>
                <source>${java.version}</source>
                <target>${java.version}</target>
                <annotationProcessorPaths>
                    <path>
                        <groupId>org.projectlombok</groupId>
                        <artifactId>lombok</artifactId>
                        <version>1.18.26</version>
                    </path>
                </annotationProcessorPaths>
            </configuration>
        </plugin>
        <plugin>
            <groupId>io.github.fvarrui</groupId>
            <artifactId>javapackager</artifactId>
            <version>1.7.3-20230613.105302-8</version>
            <configuration>
                <mainClass>claygminx.worshipppt.Main</mainClass>
                <bundleJre>true</bundleJre>
                <customizedJre>true</customizedJre>
                <url>https://github.com/ClayGminx/worship-ppt</url>
                <generateInstaller>false</generateInstaller>
                <administratorRequired>false</administratorRequired>
                <additionalResources>
                    <additionalResource>config</additionalResource>
                    <additionalResource>PPT模板</additionalResource>
                    <additionalResource>数据库</additionalResource>
                    <additionalResource>README.md</additionalResource>
                </additionalResources>
            </configuration>
            <executions>
                <execution>
                    <id>bundling-for-windows</id>
                    <phase>package</phase>
                    <goals>
                        <goal>package</goal>
                    </goals>
                    <configuration>
                        <platform>windows</platform>
                        <createZipball>true</createZipball>
                        <createTarball>false</createTarball>
                        <jrePath>jre-win</jrePath>
                        <winConfig>
                            <exeCreationTool>winrun4j</exeCreationTool>
                            <wrapJar>true</wrapJar>
                            <fileDescription>A worship PPT Tool</fileDescription>
                            <copyright>Joseph Mark</copyright>
                            <language>SIMPLIFIED_CHINESE</language>
                        </winConfig>
                    </configuration>
                </execution>
                <execution>
                    <id>bundling-for-mac</id>
                    <phase>package</phase>
                    <goals>
                        <goal>package</goal>
                    </goals>
                    <configuration>
                        <platform>mac</platform>
                        <createZipball>true</createZipball>
                        <createTarball>false</createTarball>
                        <jrePath>jre-mac</jrePath>
                        <macConfig>
                            <generateDmg>false</generateDmg>
                            <generatePkg>false</generatePkg>
                        </macConfig>
                    </configuration>
                </execution>
            </executions>
        </plugin>
    </plugins>
</build>

<pluginRepositories>
    <pluginRepository>
        <id>nexus</id>
        <name>nexus-snapshot-repository</name>
        <url>https://oss.sonatype.org/content/repositories/snapshots</url>
        <snapshots>
            <enabled>true</enabled>
            <updatePolicy>always</updatePolicy>
        </snapshots>
        <releases>
            <enabled>false</enabled>
        </releases>
    </pluginRepository>
</pluginRepositories>
```

I tried to package in Intellij IDEA ULTIMATE 2020.1, and my OS is Windows 10 家庭中文版 22H2. However, the maven package failed. The last important package log is:

```
[INFO]         Executing command: cmd.exe /s /c ""C:\Program Files\Java\jdk1.8.0_351\jre\bin\jar" uf "D:\Workspace\worship-ppt\target\Worship PPT\libs\winrun4j-launcher.jar" winrun4j.properties"
[ERROR]         '"C:\Program Files\Java\jdk1.8.0_351\jre\bin\jar"' �����ڲ����ⲿ���Ҳ���ǿ����еĳ���
[ERROR]         ���������ļ���
```

Then I tried to execute the command from the above log in cmd.exe:

```
"C:\Program Files\Java\jdk1.8.0_351\jre\bin\jar" uf "D:\Workspace\worship-ppt\target\Worship PPT\libs\winrun4j-launcher.jar" winrun4j.properties
```

But failed.

```
'"C:\Program Files\Java\jdk1.8.0_351\jre\bin\jar"' 不是内部或外部命令，也不是可运行的程序
或批处理文件。
```

**Comment by ClayGminx** [comment]
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: D:\Apps\apache-maven-3.6.3\bin\..
Java version: 1.8.0_351, vendor: Oracle Corporation, runtime: C:\Program Files\Java\jdk1.8.0_351\jre
Default locale: zh_CN, platform encoding: GBK
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"

**Comment by ClayGminx** [comment]
@fvarrui Thank you for your suggestion that I can package using **GitHub Action**, but I want to use customized jre, because the jre on Windows OS is too large.

**Comment by fvarrui** [comment]
> @fvarrui Thank you for your suggestion that I can package using **GitHub Action**, but I want to use customized jre, because the jre on Windows OS is too large.

If you want to generate a customized JRE you need a higher Java version (JDK 9+), which includes `jlink` tool.

**Comment by fvarrui** [comment]
> Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f) Maven home: D:\Apps\apache-maven-3.6.3\bin.. Java version: 1.8.0_351, vendor: Oracle Corporation, runtime: C:\Program Files\Java\jdk1.8.0_351\jre Default locale: zh_CN, platform encoding: GBK OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"

I think you should set `JAVA_HOME=C:\Program Files\Java\jdk1.8.0_351` instead of `C:\Program Files\Java\jdk1.8.0_351\jre`. Please, could you try this and give some feedback?

**Comment by ClayGminx** [comment]
My `JAVA_HOME` is `C:\Program Files\Java\jdk1.8.0_351`

**Comment by fvarrui** [comment]
> I tried to package in Intellij IDEA ULTIMATE 2020.1, and my OS is Windows 10 家庭中文版 22H2. However, the maven package failed. The last important package log is:
> 
> ```
> [INFO]         Executing command: cmd.exe /s /c ""C:\Program Files\Java\jdk1.8.0_351\jre\bin\jar" uf "D:\Workspace\worship-ppt\target\Worship PPT\libs\winrun4j-launcher.jar" winrun4j.properties"
> [ERROR]         '"C:\Program Files\Java\jdk1.8.0_351\jre\bin\jar"' �����ڲ����ⲿ���Ҳ���ǿ����еĳ���
> [ERROR]         ���������ļ���
> ```
> 
> Then I tried to execute the command from the above log in cmd.exe:
> 
> ```
> "C:\Program Files\Java\jdk1.8.0_351\jre\bin\jar" uf "D:\Workspace\worship-ppt\target\Worship PPT\libs\winrun4j-launcher.jar" winrun4j.properties
> ```
> 
> But failed.
> 
> ```
> '"C:\Program Files\Java\jdk1.8.0_351\jre\bin\jar"' 不是内部或外部命令，也不是可运行的程序
> 或批处理文件。
> ```

Hi @ClayGminx!
No news, good news? 😅 

I'm not sure what's wrong in your config, but as you can see in `mvn -v` output, Maven is using a JRE instead of a JDK:

```bash
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: D:\Apps\apache-maven-3.6.3\bin..
Java version: 1.8.0_351, vendor: Oracle Corporation, runtime: C:\Program Files\Java\jdk1.8.0_351\jre
Default locale: zh_CN, platform encoding: GBK
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"
```

When you use WinRun4J with JP, it generates a `winrun4j.properties` file, and includes this file in `winrun4j-launcher.jar` using the `jar` command, which is not included within the JRE.

`winrun4j-launcher.jar` is a work around to a WinRun4J limitation: it only can run a class without package (stored in the root package). `winrun4j-launcher.jar` includes this class which acts as a launcher for your main class. This launcher class reads `winrun4j.properties` file to know what is your main class.

But as I said before, if you want a customized JRE you have to use a JDK 9+.

**Comment by ClayGminx** [comment]
@fvarrui I'm waiting for your reply in fact. I believe my config of Java and Maven is true. Although I switch Maven version from 3.6 to 3.9, the runtime of Maven is still `C:\Program Files\Java\jdk1.8.0_351\jre`.

When I edit my reply, I try to switch my JDK version to 17 which doesn't have jre directory, and I set `JAVA_HOME` to `C:\Program Files\Java\jdk-17.0.7`, then I try the command `mvn -v`, the output is

```
Apache Maven 3.9.3 (21122926829f1ead511c958d89bd2f672198ae9f)
Maven home: D:\Apps\apache-maven-3.9.3
Java version: 17.0.7, vendor: Oracle Corporation, runtime: C:\Program Files\Java\jdk-17.0.7
Default locale: zh_CN, platform encoding: GBK
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"
```

Although, I don't want to upgrade my JDK version.

Additionally, the method of using customized JRE 8 or reducing JRE 8 is removing files about JavaFX.

**Comment by fvarrui** [comment]
> My `JAVA_HOME` is `C:\Program Files\Java\jdk1.8.0_351`

This confused me a bit, since `mvn -v` says that `C:\Program Files\Java\jdk1.8.0_351\jre` is the Java runtime. It's a bit weird, but I think I have a solution.

As I mentioned earlier, JP uses the JDK's `jar` command. The problem is that with Java 8, Maven points to your JRE (`%JAVA_HOME%\jre`), instead of what is indicated in the **JAVA_HOME** variable. I don't know why, but it is so. The `jar` command is not in the JRE, but in the JDK, hence the error.

```bash
[INFO]         Executing command: cmd.exe /s /c ""C:\Program Files\Java\jdk1.8.0_351\jre\bin\jar" uf "D:\Workspace\worship-ppt\target\Worship PPT\libs\winrun4j-launcher.jar" winrun4j.properties"
[ERROR]         '"C:\Program Files\Java\jdk1.8.0_351\jre\bin\jar"' �����ڲ����ⲿ���Ҳ���ǿ����еĳ���
[ERROR]         ���������ļ���
```

Here the code:

https://github.com/fvarrui/JavaPackager/blob/05a219554abcd55675b85fcc4b1fd127b972d3a6/src/main/java/io/github/fvarrui/javapackager/utils/JarUtils.java#L10-L13

I think that replacing line 11 with next fix the problem:

```java
File jar = new File(System.getenv("JAVA_HOME"), "/bin/jar");
```
so `jar` variable points to:
-  `C:\Program Files\Java\jdk1.8.0_351\bin\jar` (`System.getenv("JAVA_HOME")`) 
instead of 
- `C:\Program Files\Java\jdk1.8.0_351\jre\bin\jar` (`System.getProperty("java.home")`)

**Comment by fvarrui** [comment]
If you agree with me, I'll code a patch for this and publish a snapshot version, so you can test it.

**Comment by fvarrui** [comment]
Changes made in branch `issue-325`.

**Comment by ClayGminx** [comment]
Thank you! Agree.

**Comment by ClayGminx** [comment]
@fvarrui Thank you. Now when I use the snapshot version `1.7.3-20230710.112156-15`, my exe file can display Chinese text normally.
When will you publish the `1.7.3`?
![1](https://github.com/fvarrui/JavaPackager/assets/62022018/e823a3c0-30b5-4f09-b6e0-effb9027aa87)
![2](https://github.com/fvarrui/JavaPackager/assets/62022018/d2c8fa57-b664-4961-ace0-e25fbfe1abe3)

**Comment by fvarrui** [comment]
issue-325 branch merged into devel, ready to be released in v1.7.3

**Comment by fvarrui** [comment]
1.7.3 released to Maven Central
